### DISCUSSION TITLE
Problem installing x11iraf-2.1

### DISCUSSION THREAD

**Discussion by anstockton** [iraf-community/iraf#261] [Q&A] [answered]
I am having an issue installing x11iraf-2.1 (MacBook Air M1 running macOS 12.6.3. Here is the printout:
```
stockton@macbook-air x11iraf-2.1 % make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C xgterm
cc -L/usr/X11/lib -o xgterm button.o charproc.o cursor.o data.o gtermio.o input.o main.o menu.o misc.o screen.o scrollbar.o tabs.o util.o VTPrsTbl.o \
            -L../obm -lobm \
            -lXpm -ltcl -lXaw3d -lXmu -lXt -lXext -lX11 -lncurses -lm
Undefined symbols for architecture arm64:
  "_Tcl_GetErrorLine", referenced from:
      _ObmDeliverMsg in libobm.a(Obm.o)
      _obm_call_activate_callbacks in libobm.a(Obm.o)
      _gtermInputCallback in libobm.a(gterm.o)
      _gtermResizeCallback in libobm.a(gterm.o)
      _gtermResetCallback in libobm.a(gterm.o)
      _cb_error in libobm.a(html.o)
      _MarkerEvaluate in libobm.a(marker.o)
      ...
  "_Tcl_SetErrorLine", referenced from:
      _MarkerEvaluate in libobm.a(marker.o)
      _ParameterEvaluate in libobm.a(param.o)
      _WidgetEvaluate in libobm.a(widget.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [xgterm] Error 1
make: *** [xgterm] Error 2
```

The link for .iraf/bin/xc looks correct, and $IRAF points to the iraf-2.17 directory. It seems to be a problem finding the correct library files. Any help would be appreciated. Thanks.

**Comment by olebole** [comment, marked as answer]
This is probably the same problem as in #228 resp. iraf-community/x11iraf#46: for some reason, the "brew" installs a newer `tcl.h` on `/usr/local/include`, where it is found before the system's one `/usr/include/tcl.h`. This newer header relies on a newer libtcl, which is however not in the path. There are two ways to fix this: Either you remove `/usr/local/include/tcl.h` (or move if somewhere else), or add the path of the corresponding libtcl to `LD_LIBRARY_PATH`. This is a bug in brew, which should be reported there.

**Reply by anstockton** [reply]
I tried removing /usr/local/include/tcl.h, which changed the output from "make," but I am still having some problem:

stockton@macbook-air x11iraf-2.1 % make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C xgterm
make[1]: Nothing to be done for `all'.
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ximtool
cd clients && mkpkg || true
/bin/sh: mkpkg: command not found

**Reply by olebole** [reply]
This is probably the same problem as in #228: The mkpkg program (which is part of IRAF) must be in the PATH. It is however only needed if you want to use ximtool and need the WCS coordinated displayed in the tool. If you use ds9 as image display, this can safely ignored.

**Reply by anstockton** [reply]
Sorry to keep bothering on this. I did find the reason that it wasn't getting to mkpkg: I had mistyped the directory path for the IRAF distribution when putting it into $PATH (/User/...instead of /Users/...). But I also noticed that the permission for mkpkg did not allow execution (644 instead of 755), so I changed that. But it is still not working properly. As you guessed, I really don't need ximtool, but I am not getting a working xgterm, which I do need. Here is the latest output:

[stockton@macbook-air x11iraf-2.1 % make
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C xgterm
make[1]: Nothing to be done for `all'.
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ximtool
cd clients && mkpkg || true
/Users/stockton/iraf-2.17/mkpkg: line 5: mkall: command not found
/Users/stockton/iraf-2.17/mkpkg: line 8: sysgen:: command not found
/Users/stockton/iraf-2.17/mkpkg: line 9: vendor: command not found
/Users/stockton/iraf-2.17/mkpkg: line 10: mkall: command not found
/Users/stockton/iraf-2.17/mkpkg: line 11: update: command not found
/Users/stockton/iraf-2.17/mkpkg: line 12: syntax error near unexpected token `;'
/Users/stockton/iraf-2.17/mkpkg: line 12: `	;'

Maybe I should start all over with a completely new IRAF2.17 install, then have another go at X11IRAF, incase something has gotten corrupted?

**Reply by olebole** [reply]
Do you need ximtool? With WCS coordinates? As I wrote, for xgterm this is not needed.

**Reply by anstockton** [reply]
No; as I said, I don't need ximtool, but I cannot get xgterm working either.

**Reply by olebole** [reply]
why not? xgterm is built. Put it somewhere in your path and it should work. What happens if you start it?

**Comment by anstockton** [comment]
Sorry, stupid confusion between the directory xgterm and the executable
within. Thanks; you may close this now.

On Fri, Mar 3, 2023 at 10:17 AM Ole Streicher ***@***.***>
wrote:

> why not? xgterm is built. Put it somewhere in your path and it should
> work. What happens if you start it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/iraf-community/discussions/261#discussioncomment-5196689>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO5I77RNYCIP3A74Y6P2JTW2JGTJANCNFSM6AAAAAAVN3ZZJU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Alan Stockton
Institute for Astronomy
University of Hawaii
2680 Woodlawn Drive
Honolulu, Hawaii 96822

**Reply by olebole** [reply]
One final remark: IRAF is really old software, nowadays lacks further development and and institutional support. It is mainly maintained for those who cannot switch yet. Unless there is a real strong requirement, it is a really bad idea to start with it for a new project. It would be much better to use modern Python projects like [Astropy](https://astropy.org/) instead whenever it is suitable. This is much more future-proof than learning an outdated, 40 year old legacy software system.

**Comment by anstockton** [comment]
That is certainly good advice in general. But I am 80 years old, have been
retired for a number of years, and have been using IRAF ever since
it began. I thought that I would try to finish up one last project, and I
have quite quite a bit invested in custom IRAF reduction scripts. I am not
sure that they will all still work with the current version, but I suspect
that I can work around that problem to get what I need more efficiently
than trying to convert everything to Python. Thanks again for your help.

On Fri, Mar 3, 2023 at 11:59 AM Ole Streicher ***@***.***>
wrote:

> One final remark: IRAF is really old software, nowadays lacks further
> development and and institutional support. It is mainly maintained for
> those who cannot switch yet. Unless there is a real strong requirement, it
> is a really bad idea to start with it for a new project. It would be much
> better to use modern Python projects like Astropy <https://astropy.org/>
> instead whenever it is suitable. This is much more future-proof than
> learning an outdated, 40 year old legacy software system.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/iraf-community/discussions/261#discussioncomment-5197420>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO5I76IEE6K423FLX2XWYDW2JSSLANCNFSM6AAAAAAVN3ZZJU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Alan Stockton
Institute for Astronomy
University of Hawaii
2680 Woodlawn Drive
Honolulu, Hawaii 96822

**Reply by olebole** [reply]
That is ofcourse a good reason! :+1:

**Reply by anstockton** [reply]
One more problem:

stockton@macbook-air iraf % ecl
setting terminal type to 'xgterm' ...
ERROR: Cannot open connected subprocess (./x_system.e)
     called as: `cl ()'
Error while reading login.cl file - may need to rebuild with mkiraf
Fatal startup error.  CL dies.
Warning: Attempt to delete a nonexistent file (uparm$pipe4148)

From what I understand, the x_system.e file should be in iraf-2.17/bin.macos64, where the cl.e and ecl.e are located, but the distribution I have does not include this file:
...
/Users/stockton/iraf-2.17/bin.macos64/x_sgikern.e
/Users/stockton/iraf-2.17/bin.macos64/x_softools.e
/Users/stockton/iraf-2.17/bin.macos64/x_stdgraph.e
/Users/stockton/iraf-2.17/bin.macos64/x_utilities.e

**Reply by olebole** [reply]
Did you compile IRAF before? The distribution contains only the sources; you first need to compile. If you did, and `x_system.e` is not in `bin.macos64`, then you need to look into the log file `make.log` for compilation errors.

**Reply by anstockton** [reply]
I seem to be going backwards. I tried reinstalling IRAF-2.17 from the directions several times, but now after running ./install and make macos64, there is nothing at all in the bin.macos64 directory or the unix/bin.macos64 directory. I have /Users/stockton/.iraf/ in my $PATH, and the variables $IRAFARCH and $iraf are set correctly. But of course, since the bin.macos64 directories are empty, xc cannot be found. Any ideas on what could cause this behavior?

**Reply by anstockton** [reply]
And when I run "make sysgen 2>&1 | tee build.log" on a previous version that seemed to go successfully through the above steps, so that xc is available, it gives the following:

stockton@macbook-air iraf-2.17 % make sysgen 2>&1 | tee build.log
Building the IRAF v2.16 software tree
 
Mon Mar  6 14:12:50 HST 2023
 
(util/mksysgen)
Clean 'macos64' host directories ....
Clean src directories ....
Clean 'macos64' bin directories ....
Cleaning spool files ....
+ set -e
+ sh -x setarch.sh
++ ls -d bin
+ '[' bin = bin ']'
+ rm -rf bin
+ ln -s bin.macos64 bin
+ echo '----------------------- OS -----------------------------'
----------------------- OS -----------------------------
+ echo +
+
+ echo +
+
+ cd os
+ sh -x mkpkg.sh
+ set -e
+ echo '--------------------- OS ----------------------'
--------------------- OS ----------------------
+ gcc -c -g -Wall -O2 -I/Users/stockton/iraf-2.17/include -DNOVOS -Wall zsvjmp.S gmttolst.c irafpath.c prwait.c zalloc.c zawset.c zcall.c zdojmp.c zfacss.c zfaloc.c zfchdr.c zfdele.c zfgcwd.c zfinfo.c zfiobf.c zfioks.c zfiolp.c zfiomt.c zfiond.c zfiopl.c zfiopr.c zfiosf.c zfiotx.c zfioty.c zflink.c zfmkcp.c zfmkdr.c zfnbrk.c zfpath.c zfpoll.c zfprot.c zfrmdr.c zfrnam.c zfsubd.c zfunc.c zfutim.c zfxdir.c zgcmdl.c zghost.c zglobl.c zgmtco.c zgtenv.c zgtime.c zgtpid.c zintpr.c zlocpr.c zlocva.c zmain.c zmaloc.c zmfree.c zopdir.c zopdpr.c zoscmd.c zpanic.c zraloc.c zshlib.c zttyio.c zwmsec.c zxwhen.c zzdbg.c zzepro.c zzexit.c zzpstr.c zzsetk.c zzstrt.c alloc.c getproc.c
mkpkg.sh: line 10: 14090 Bus error: 10           $CC -c $HSI_CF -Wall zsvjmp.S gmttolst.c irafpath.c prwait.c z*.c alloc.c getproc.c
make: *** [sysgen] Error 138

**Reply by olebole** [reply]
It is correct that after `make macos64` the `bin.macos64` directories remain empty; they are filled only in the earlier stages of `make sysgen`. and you should append `~/.iraf/bin` to your PATH, not just `~/.iraf`.

The bus error is surely a problem of the C compiler or the hardware, not of IRAF. Make sure that you use the compiler delivered with macOS and not something that comes f.e. with brew or macPorts.

**Reply by anstockton** [reply]
Sorry, I did have _/.iraf/bin in my PATH but missed putting the "bin" in my message. Thanks for the hint about the compiler–I found that it was trying to use an old gcc in /usr/local/bin instead of the one in /usr/bin. The make sysgen now goes all of the way, but there are a lot of warnings, some of which look to be inconsequential, but quite a few seem important, e.g., not being able to find libsys.a, libpkg.a, etc. (even though at least the first is in iraf_2.17/lib), and a bunch of warnings on mkpkg not being able to find modules such as dtoi, relink, install, update@dtoi. There are more.

**Reply by olebole** [reply]
You should have a look into the build log; probably there is a line `xc: Command not found` or so. In this case, you should look in the build log that `xc.e` was built properly and moved to the right place. Like:
```
----------------------- XC  ----------------------------
+ gcc -c -g -Wall -O2 -I/usr/local/iraf/iraf-2.17/include -DNOVOS xc.c
+ gcc xc.o /Users/icke/iraf/iraf-2.17/unix/hlib/libboot.a /Users/icke/iraf/iraf-2.17/unix/hlib/libos.a
 -lm -o xc.e
+ mv -f xc.e ../../hlib
+ rm -f xc.o
[…]
+ echo 'install HSI executables in /Users/icke/iraf/iraf-2.17/unix//bin.macos64'
install HSI executables in /Users/icke/iraf/iraf-2.17/unix//bin.macos64
+ mv -f hlib/alloc.e hlib/generic.e […] hlib/xc.e hlib/xpp.e hlib/xyacc.e bin.macos64
```
And you should make sure that the `$iraf` environment variable does not point to a different place.

**Reply by anstockton** [reply]
Sorry to bother again, but there is no problem with xc; it has been moved correctly to ~/iraf-2.17/unix/bin, and the link in .iraf/bin points to it. I am attaching two text files with all of the warnings for "not found" and "no such file or directory" from the build.log.
[not_found.txt](https://github.com/iraf-community/iraf/files/10914402/not_found.txt)
[no_such.txt](https://github.com/iraf-community/iraf/files/10914406/no_such.txt)

**Reply by olebole** [reply]
I just made some changes in the code to make the build process more robust. Maybe you could try out this test version:

https://github.com/olebole/iraf/archive/refs/heads/mkpkg-vars.tar.gz

Building on a pristine Mac succeeded in a test. If it fails on your computer, please attach the build log.
For discussion, this is Pull Request https://github.com/iraf-community/iraf/pull/262.

**Comment by anstockton** [comment]
I'm afraid that it looks like it didn't make any difference. I suspect that
the problem may be stuff that I ported over from an older Mac, but I don't
know where the issue might be. I did clean out all of the old stuff from
/usr/local/bin. Here is the build.log:

On Thu, Mar 9, 2023 at 7:03 AM Ole Streicher ***@***.***>
wrote:

> I just made some changes in the code to make the build process more
> robust. Maybe you could try out this test version:
>
> https://github.com/olebole/iraf/archive/refs/heads/mkpkg-vars.tar.gz
>
> Building on a pristine Mac succeeded in a test. If it fails on your
> computer, please attach the build log.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/iraf-community/discussions/261#discussioncomment-5257025>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACO5I76UJEI4NY2ZZXWIUD3W3IEKVANCNFSM6AAAAAAVN3ZZJU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Alan Stockton
Institute for Astronomy
University of Hawaii
2680 Woodlawn Drive
Honolulu, Hawaii 96822

**Reply by olebole** [reply]
The build log is not there. You should not attach it to the mail, but put it on the discussion web site instead.

**Reply by anstockton** [reply]
Sorry–should have guessed that. Here it is.

[build.log](https://github.com/iraf-community/iraf/files/10939464/build.log)

**Reply by olebole** [reply]
The reason is here:
```
…
--------------------- SYSTEM ----------------------
warning: library `libpkg.a' not found
/Users/stockton/iraf-mkpkg-vars/unix/bin.macos64/xc.e -c -w breakline.x center.x dols.x getarg.x …
Termination code 11ar r /Users/stockton/iraf-mkpkg-vars/pkg/system/libpkg.a breakline.o center.o …
```
This shows, that now the correct "xc" command is found now. However, it looks like that probably one of the programs xc calls (C compiler?) doesn't work well and causes a segmentation fault ("Termination code 11"). Are you sure that you use the C compiler that comes with macOS/XCode and not something else?

You could go to the corresponding directory (`pkg/system/help/lroff /`) and try to run the command manually; with an added `-d` option:
```
/Users/stockton/iraf-mkpkg-vars/unix/bin.macos64/xc.e -d -c -w breakline.x center.x dols.x getarg.x 
```
This print all commands that are called by xc, helping to find which one crashes.

**Reply by anstockton** [reply]
Here is the output from the manual run of the command:
```
stockton@macbook-air lroff % /Users/stockton/iraf-2.17/unix/bin.macos64/xc.e -d -c -w breakline.x center.x dols.x getarg.x
debug: /Users/stockton/iraf-2.17//unix/bin.macos64/xpp.e -R breakline.x
debug: /Users/stockton/iraf-2.17//unix/bin.macos64/rpp.e breakline.r >breakline.f
debug: /Users/stockton/iraf-2.17//unix/bin.macos64/xpp.e -R center.x
debug: /Users/stockton/iraf-2.17//unix/bin.macos64/rpp.e center.r >center.f
debug: /Users/stockton/iraf-2.17//unix/bin.macos64/xpp.e -R dols.x
debug: /Users/stockton/iraf-2.17//unix/bin.macos64/rpp.e dols.r >dols.f
debug: /Users/stockton/iraf-2.17//unix/bin.macos64/xpp.e -R getarg.x
debug: /Users/stockton/iraf-2.17//unix/bin.macos64/rpp.e getarg.r >getarg.f
/Users/stockton/iraf-2.17//unix/hlib/f77.sh -c -u -x -f2c /Users/stockton/iraf-2.17//unix/bin.macos64/f2c.e -O2 -w -I/Users/stockton/iraf-2.17//unix/hlib/libc breakline.f center.f dols.f getarg.f
```
So I gather that it is hanging on the f77.sh script. As far as I can tell from trying to run this manually, it is failing in the first do loop atz;

	-m*)	CFLAGS="$CFLAGS $1"
		shift 1
		;;

Though why it should fail there is beyond my understanding.

**Reply by olebole** [reply]
Did you get the "Termination code 11" here?

The code that you have shown is part of the shell syntax and can't be executed on its own. f77.sh executes two programs that may be the cause: `/Users/stockton/iraf-2.17/unix/bin.macos64/f2c.e` and the "gcc" C compiler that is in your PATH. Are you sure that you use the clang C compiler that comes with the system, or maybe a different one that doesn't work properly?

**Reply by olebole** [reply]
I now added some more changes so that one can set the C compiler by command line. Could you download  https://github.com/olebole/iraf/archive/refs/heads/mkpkg-vars.tar.gz again, and replace the `make sysgen` line with this one?
```shell
CC=/usr/bin/clang make sysgen 2>&1 | tee build.log
```
This will make sure that the build process uses the default macOS/XCode compiler.

**Reply by anstockton** [reply]
I ran it with this command. Fewer errors, I think, but still a number of problems, especially with various libxxx.a. I attach the log, but let me play around with it tomorrow before you send me anything else. I am pretty sure that the problems have to do with stuff ported over from a previous Mac running 32-bit code. I am going to try to clean out all of that and try again.
[make.log](https://github.com/iraf-community/iraf/files/10950329/make.log)

**Reply by olebole** [reply]
You should re-download the tar file again, since I updated it yesterday to correctly recognize the `CC` environment variable.

**Reply by anstockton** [reply]
Re-downloaded the iraf-mkpkg-vars.tar.gz and ran through the install procedures using CC=/usr/bin/clang make sysgen 2>&1 | tee build.log. I also deleted virtually all of the old files imported from the 32-bit Mac in /usr/local. Still no dice; here is the build log:
[build.log](https://github.com/iraf-community/iraf/files/10954770/build.log)

I don't want to waste any more of your time on this, to tomorrow I plan to do a clean install of Ventura and reinstall Xcode. It will probably work then, I should hope.

**Reply by anstockton** [reply]
I have finally done a clean install of 13.2.1 and am preparing to try the IRAF install again. I note changes in the install procedure, including reference to 2.17.1. But the GitHub page still gives the tar.gz download as 2.17. Where do I find iraf-2.17.1.tar.gz?

**Reply by olebole** [reply]
The version I gave you is a development version; there is no 2.17.1. yet.

**Reply by anstockton** [reply]
OK, I'm confused then, since the current instructions show the following:
~~~~~~~~~~
Unpack the IRAF Distribution
The source distribution file is built as a tarball with the package name and version as base directory. Thus, distribution files can be unpacked with the command

$ tar zxf /<path>/iraf-2.17.1.tar.gz
$ cd iraf-2.17.1/
Build from Sources
Now you can compile IRAF on your system with the command

$ make 2>&1 | tee build.log
~~~~~~~~~~
Also eliminating the "make macos64" step. How does it know which architecture is it being compiled for?

**Reply by olebole** [reply]
It retrieves it automatically with `unix/hlib/irafarch`. The idea is to further simplify the installation process. For special cases (like compiling 32-bit on a 64-bin Linux) one still can do this step.

**Reply by anstockton** [reply]
Still no luck, even with the new install and using your iraf-mkpkg-vars and CC=/usr/bin/clang make sysgen 2>&1 | tee build.log. Here is the build.log:
[build.log](https://github.com/iraf-community/iraf/files/11009937/build.log)

I do have an old laptop for which the MacOS has not been upgraded, so it can still run 2.16.1; going back to that may be my best option for this one project.
